src/omotes_simulator_core/entities/network_controller.py

```diff
@@ -44,6 +44,23 @@ def __init__(
         self.producers = producers
         self.consumers = consumers
         self.storages = storages
+        self._set_priority_from_marginal_costs()
+
+    def _set_priority_from_marginal_costs(self) -> None:
+        """This method sets the priority of the producers based on the marginal costs."""
+        # Sort the producers based on the marginal costs
+        self.producers.sort(key=lambda x: x.marginal_costs)
+
+        # Set the priority based on the sorted list
+        previous_marginal_costs = self.producers[0].marginal_costs
```

Comment:
It probably doesn't help but is there a way to remove the nested loops? 
```suggestion
        a = [15, 0, 10, 20, 0, 10, 15, 15]

        # Sort the list and remove duplicates
        unique_sorted_values = sorted(set(a))

        # Assign an index to each unique value
        indexed_values = {value: index for index, value in enumerate(unique_sorted_values)}

```

Reply:
There is no nested loop. The loop over the producers is only called once. 
What you could to make a sorted list of the marginal costs and then set priority on the index + 1 in this lost of the marginal costs.
I adjusted it, so code becomes more cleaner and smaller.